### DISCUSSION TITLE
Programmatically set value and trigger a search

### DISCUSSION THREAD

**Discussion by Riddler84** [orchidjs/tom-select#240] [Q&A]
Hello,
I may have a somewhat unusual use case and would like to know if this is even possible. I have tried a few functions and triggers now, but nothing has worked.

I would like to include two select fields, each making AJAX requests to different sources. I would like to enter the search word via another text field and this should then be transferred to the two select fields, which then perform a search if possible without opening the dropdown and automatically select the first result.

Using "setTextboxValue()" I get the value into the input fields of the select fields, but nothing else happens. I also tried to trigger input events using "trigger()", but also without success.

A push in the right direction would be nice^^

**Comment by mholt** [comment]
I would also like to auto-populate the select with some values from the query string when the page loads. Since the select gets the data remotely, it would have to perform a search and select the top result (I know the top result will be correct because it's searching by item ID).

How can this be done?

Thanks for any help!

**Comment by mholt** [comment]
I think I figured out a way.

```
control.addOption({id: 15});
control.addItem(15);
```

It's close to an example in the docs, but it wasn't very obvious to me. Basically, the argument to `addOption` is the same type of value you search over. Then you use the "value" of the data you just passed in to select it.

**Reply by shifuma** [reply]
Could I please get you to give a more complete example of how you're using this? (sorry, I know this is old)

**Reply by mholt** [reply]
That's literally all I do for this. I have my object, passed into `addOption()`, and its ID (which is the value used for the form), passed into `addItem()`.

I don't have a complete example setting up the TomSelect that I can just paste in here because I've refactored my code so it can be flexible enough for my growing web app.

**Reply by shifuma** [reply]
Thanks for the reply. This is my first time using Tom Select...where did you add `control.addOption()`?

**Reply by mholt** [reply]
What does your code look like so far?

**Comment by incredibleasif** [comment]
Hi 👋

I had a very similar requirement where:

* Options must load via **AJAX**
* Search keyword comes from **another input field**
* Dropdown should **not open**
* First result should be **auto-selected**
* Second dropdown depends on first dropdown value

After some testing, this approach worked for me.

I’m using **CodeIgniter 4 (backend)** + **Tom-Select with jQuery (frontend)**.

---

### Working Solution

```javascript
const selectedVal   = "{Selected Value for this dropdown}";
const parentFieldId = "parentField";

const parentField = new TomSelect('#' + parentFieldId, {
    maxItems: 1,
    valueField: 'value',
    labelField: 'text',
    searchField: ['value','text'],

    load: async function(query, callback) {
        const formData = new FormData();
        formData.append('q', query);

        try {
            const response = await fetch("/api/get-items", {
                method: 'POST',
                body: formData,
                headers: {
                    'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content') || ''
                }
            });

            const json = await response.json();
            callback(json.data);

        } catch (e) {
            callback();
        }
    },

    create: false,
});
```

---

### Trigger search programmatically & auto-select first result

```javascript
// After results load → select value automatically
parentField.on('load', function () {
    parentField.setValue(selectedVal);
    parentField.off('load'); // run only once on page load
});

// Perform search using external value (without opening dropdown)
parentField.load(selectedVal);

```

---

### Child Dropdown (depends on parent value)

Now you can initialize the **child field** and pass parent value in AJAX:

```javascript
load: async function(query, callback) {
    const formData = new FormData();
    formData.append('q', query);

// Send parent selected value
    formData.append('parent_item', parentField.getValue());

    try {
        const response = await fetch("/api/get-child-items", {
            method: 'POST',
            body: formData,
            headers: {
                'X-CSRF-TOKEN': $('meta[name="csrf-token"]').attr('content') || ''
            }
        });

        const json = await response.json();
        callback(json.data);

    } catch (e) {
        callback();
    }
}
```

---

### Notes

* `load(value)` can be used to **force search programmatically**
* Use `on('load')` to auto-select once results arrive
* Works without opening dropdown
* Useful for **auto-populate + dependent dropdown via AJAX**

Hope this helps 👍
